pkg/reconciler/v1alpha1/taskrun/resources/entrypoint.go

```diff
@@ -0,0 +1,142 @@
+// copy the entrypoint binary from the entrypoint image into the
+// volume mounted at mountPoint, so that it can be mounted by
+// subsequent steps and used to capture logs.
+func GetCopyStep() corev1.Container {
+	return corev1.Container{
+		Name:         "place-tools",
+		Image:        EntrypointImage,
+		Command:      []string{"/bin/cp"},
+		Args:         []string{"/entrypoint", entrypointBin},
+		VolumeMounts: []corev1.VolumeMount{toolsMount},
+	}
+}
+
+type entrypointArgs struct {
+	Args       []string `json:"args"`
+	ProcessLog string   `json:"process_log"`
+	MarkerFile string   `json:"marker_file"`
+}
+
+func getEnvVar(baseCmd, cmd, args []string) (string, error) {
+	// use the default entrypoint for the container
+	as := append(baseCmd, args...)
+	if len(cmd) > 0 {
+		// unless the step actually overrides the entrypoint
+		as = append(cmd, args...)
+	}
+	entrypointArgs := entrypointArgs{
+		Args:       as,
+		ProcessLog: "/tools/process-log.txt",
+		MarkerFile: "/tools/marker-file.txt",
+	}
+	j, err := json.Marshal(entrypointArgs)
+	if err != nil {
+		return "", fmt.Errorf("couldn't marshal arguments %q for entrypoint env var: %s", entrypointArgs, err)
+	}
+	return string(j), nil
+}
+
+// TODO: add more test cases after all, e.g. with existing env
+// var and volume mounts
+
+// AddEntrypoint will modify each of the steps/containers such that
+// the binary being run is no longer the one specified by the Command
+// and the Args, but is instead the entrypoint binary, which will
+// itself invoke the Command and Args, but also capture logs.
+// TODO: This will not work when a step uses an image that has its
+// own entrypoint, i.e. `Command` is a required field. In later iterations
+// we can update the controller to inspect the image's `Entrypoint`
+// and use that if required.
+func AddEntrypoint(steps []corev1.Container) error {
+	cache := make(map[string][]string)
+	getEntrypoint := func(image string) ([]string, error) {
+		// check if we've seen this image before
+		if ep, ok := cache[image]; ok {
+			return ep, nil
+		}
+		// verify the image name, then download the remote config file
+		ref, err := name.ParseReference(image, name.WeakValidation)
+		if err != nil {
+			return nil, fmt.Errorf("couldn't parse image %s: %v", image, err)
+		}
+		img, err := remote.Image(ref, remote.WithAuthFromKeychain(authn.DefaultKeychain))
+		if err != nil {
+			return nil, fmt.Errorf("couldn't get remote image info for %s: %v", image, err)
+		}
+		cfg, err := img.ConfigFile()
+		if err != nil {
+			return nil, fmt.Errorf("couldn't get config for image %s: %v", image, err)
+		}
+
+		// add the image to the cache
+		cache[image] = cfg.ContainerConfig.Entrypoint
+		return cfg.ContainerConfig.Entrypoint, nil
+	}
+
+	for i := range steps {
+		step := &steps[i]
+		ep, err := getEntrypoint(step.Image)
+		if err != nil {
+			return fmt.Errorf("couldn't get entrypoint for %s: %v", step.Image, err)
+		}
```

Comment:
my feeling is that if there is an error here we should log it but continue onward